File: Sorting/Merge Sort/MergeSort.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> merge(vector<int> arrayA, vector<int> arrayB)
{
    vector<int> arrayC;

    while (arrayA.size() != 0 && arrayB.size() != 0)
    {
        if (arrayA[0] < arrayB[0])
        {
            arrayC.push_back(arrayA[0]);
            arrayA.erase(arrayA.begin());
        }
        else
        {
            arrayC.push_back(arrayB[0]);
            arrayB.erase(arrayB.begin());
        }
    }

    while (arrayA.size() != 0)
    {
        arrayC.push_back(arrayA[0]);
        arrayA.erase(arrayA.begin());
    }
    while (arrayB.size() != 0)
    {
        arrayC.push_back(arrayB[0]);
        arrayB.erase(arrayB.begin());
    }

    return arrayC;
}

vector<int> mergeSort(vector<int> &nums)
{
    if (nums.size() == 1)
        return nums;

    vector<int> arrayA;
    vector<int> arrayB;

    for (unsigned int i = 0; i < nums.size(); i++)
    {
        if (i < nums.size() / 2)
            arrayA.push_back(nums[i]);
        if (i >= nums.size() / 2)
            arrayB.push_back(nums[i]);
    }

    arrayA = mergeSort(arrayA);
    arrayB = mergeSort(arrayB);

    return merge(arrayA, arrayB);
}

int main()
{
    vector<int> numbers = {7, 3, 2, 5, 6, 10, 9, 8, 1};

    for (unsigned int i = 0; i < numbers.size(); i++)
    {
        cout << numbers[i] << " ";
    }
    cout << endl;

    numbers = mergeSort(numbers);

    for (unsigned int i = 0; i < numbers.size(); i++)
    {
        cout << numbers[i] << " ";
    }
    cout << endl;

    return 0;
}

File: Sorting/Merge Sort/MergeSort.js
function merge(arrayA, arrayB) {
    let arrayC = [];
    while (arrayA.length != 0 && arrayB.length != 0) {
        if (arrayA[0] < arrayB[0]) {
            arrayC.push(arrayA.shift());
        } else {
            arrayC.push(arrayB.shift());
        }
    }

    // Either A or B is empty at this point

    if (arrayA.length != 0) {
        arrayC = [...arrayC, ...arrayA];
    }
    if (arrayB.length != 0) {
        arrayC = [...arrayC, ...arrayB];
    }

    return arrayC;
}

function mergeSort(array) {
    if (array.length == 1) return array;

    let arrayA = [];
    let arrayB = [];

    for (let i = 0; i < array.length; i++) {
        if (i < array.length / 2) arrayA.push(array[i]);
        if (i >= array.length / 2) arrayB.push(array[i]);
    }

    arrayA = mergeSort(arrayA);
    arrayB = mergeSort(arrayB);

    return merge(arrayA, arrayB);
}

let nums = [7, 3, 2, 5, 6, 10, 9, 8, 1];
console.log(mergeSort(nums));
